src/sandstorm/supervisor.c++

```diff
@@ -1535,10 +1539,123 @@ void SupervisorMain::DefaultSystemConnector::checkIfAlreadyRunning() const {
 
   KJ_SYSCALL(execve(argvp[0], argvp, envp), argvp[0]);
   KJ_UNREACHABLE;
+};
+
+static void decrementWakelock() {
+  --sandstorm::wakelockCount;
+  if (sandstorm::wakelockCount == 0) {
+    SANDSTORM_LOG("Grain's backgrounding has been disabled; staying up for now.");
+    // Stay alive for one more keepAlive tick after disabling backgrounding.
+    sandstorm::keepAlive = true;
+  }
 }
 
-class SupervisorMain::SandstormApiImpl final: public SandstormApi<>::Server {
+class WakeLockInfo {
 public:
+  OngoingNotification::Client ongoingNotification;
+
+  WakeLockInfo(OngoingNotification::Client& ongoingNotification)
+    : ongoingNotification(ongoingNotification) {}
+  WakeLockInfo(WakeLockInfo&&) = default;
+  KJ_DISALLOW_COPY(WakeLockInfo);
+};
+
+class WakelockSet: private kj::TaskSet::ErrorHandler {
+public:
+
+  class WrappedOngoingNotification final: public PersistentOngoingNotification::Server {
+  public:
+    WrappedOngoingNotification(OngoingNotification::Client ongoingNotification,
+                               WakelockSet& wakelockSet)
+      : ongoingNotification(ongoingNotification), wakelockSet(wakelockSet), isCancelled(false) {
+      ++sandstorm::wakelockCount;
+    }
+    WrappedOngoingNotification(WrappedOngoingNotification&&) = delete;
+    KJ_DISALLOW_COPY(WrappedOngoingNotification);
+
+    ~WrappedOngoingNotification() {
```

Comment:
Make sure your destructors are always declared `noexcept(false)`, because C++11 rules unfortunately made them `noexcept` by default.

(Of course, you still need to make sure your destructor can't throw during unwind.)